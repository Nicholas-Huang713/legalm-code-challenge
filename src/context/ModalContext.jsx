import { createContext, useState, useEffect } from "react";
import { useDispatch } from "react-redux";
import { deleteOwner } from "../state/slices/owner/ownerThunks";
import { fetchDog } from "../state/slices/dog/dogSlice";
export const ModalContext = createContext();

export const ModalProvider = ({ children }) => {
  const dispatch = useDispatch();
  const [detailModalOpen, setDetailModalOpen] = useState(false);
  const [newOwnerModalOpen, setNewOwnerModalOpen] = useState(false);
  const [editModalOpen, setEditModalOpen] = useState(false);
  const [currentModalType, setCurrentModalType] = useState("");

  const handleCloseNewOwnerModal = () => {
    setNewOwnerModalOpen(false);
    setCurrentModalType("");
  };

  const handleCloseEditOwnerModal = () => {
    setEditModalOpen(false);
    setCurrentModalType("");
  };

  const handleCloseDetailModal = () => {
    setDetailModalOpen(false);
    setCurrentModalType("");
  };

  const handleViewDetails = async (dogId) => {
    await dispatch(fetchDog(dogId));
    setCurrentModalType("detail");
  };

  const handleOpenEditModal = async (dogId) => {
    await dispatch(fetchDog(dogId));
    setCurrentModalType("edit");
  };

  const handleDeleteOwner = async (ownerId) => {
    await dispatch(deleteOwner(ownerId));
  };

  useEffect(() => {
    if (currentModalType === "") return;
    if (currentModalType === "detail") setDetailModalOpen(true);
    else setEditModalOpen(true);
  }, [currentModalType]);

  return (
    <ModalContext.Provider
      value={{
        handleDeleteOwner,
        handleOpenEditModal,
        handleViewDetails,
        handleCloseDetailModal,
        handleCloseEditOwnerModal,
        handleCloseNewOwnerModal,
        detailModalOpen,
        newOwnerModalOpen,
        setNewOwnerModalOpen,
        editModalOpen,
      }}
    >
      {children}
    </ModalContext.Provider>
  );
};
