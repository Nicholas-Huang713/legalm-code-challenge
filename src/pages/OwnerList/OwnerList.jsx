import React from 'react'
import List from '../../components/List/List';
import {useSelector} from 'react-redux';


export default function OwnerList() {
    const ownerList = useSelector((state) => state.owner.owners);

    const handleViewDetails = (ownerId) => {
        console.log("View More Called")

    }   
    const handleEditOwner = (ownerId) => {
        console.log("Edit Called")

    }   
    const handleDeleteOwner = (ownerId) => {
        console.log("Delete Called")
    }   

    const handleBtnClick = (event) => {
        const btnType = event.target.textContent;
        const ownerId = event.target.value;
        switch (btnType) {
            case "More":
                handleViewDetails(ownerId)
                break;
            case "Edit":
                handleEditOwner(ownerId)
              break;
            case "Delete":
                handleDeleteOwner(ownerId)
              break;
            default:
          }

    }

    return (
        <>
            {ownerList.length > 0 ?
                <List 
                    list={ownerList} 
                    showBtns={true} 
                    handleBtnClick={handleBtnClick}
                />
                :null
            }
        </>
    )
}
