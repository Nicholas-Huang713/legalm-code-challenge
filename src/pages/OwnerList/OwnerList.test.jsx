import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import OwnerList from "./OwnerList";
import { ModalProvider } from "../../context/ModalContext";

import { dogDataInStore, ownersList } from "../../test/mocks/mockValues";
const thunkMiddleware = require("redux-thunk").thunk;
const mockStore = configureMockStore([thunkMiddleware]);

const renderOwnerList = () => {
  return render(
    <Provider store={store}>
      <ModalProvider>
        <OwnerList />
      </ModalProvider>
    </Provider>
  );
};

let store;

const initStoreWithDogData = () => {
  store = mockStore({
    owner: ownersList,
    dog: dogDataInStore,
  });
  store.dispatch = vi.fn();
};

beforeEach(() => {
  initStoreWithDogData();
});

describe("OwnerList Component", () => {
  it("renders List component with correct props when ownerList is not empty", () => {
    const { container, getByText } = renderOwnerList();
    expect(getByText("John Doe")).toBeInTheDocument();
    expect(getByText("Jane Doe")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('calls handleViewDetails when "Details" button is clicked', () => {
    renderOwnerList();
    const detailsButton = screen.getAllByText("Details")[0];
    fireEvent.click(detailsButton);
    expect(store.dispatch).toHaveBeenCalled();
  });

  it('calls handleEditOwner when "Edit" button is clicked', () => {
    renderOwnerList();
    fireEvent.click(screen.getAllByText("Edit")[0]);
    expect(store.dispatch).toHaveBeenCalled();
  });

  it('calls handleDeleteOwner when "X" button is clicked', () => {
    renderOwnerList();
    fireEvent.click(screen.getAllByText("X")[0]);
    expect(store.dispatch).toHaveBeenCalled();
  });

  it("filters list based on filter selection", () => {
    renderOwnerList();
    const selectBox = screen.getByLabelText(/Filter by Exp/i);
    fireEvent.change(selectBox, { target: { value: "5" } });
    expect(screen.queryByText("John Doe")).not.toBeInTheDocument();
    expect(screen.getByText("Jane Doe")).toBeInTheDocument();
    fireEvent.change(selectBox, { target: { value: "3" } });
    expect(screen.queryByText("Jane Doe")).not.toBeInTheDocument();
    expect(screen.getByText("John Doe")).toBeInTheDocument();
    fireEvent.change(selectBox, { target: { value: "none" } });
    expect(screen.getAllByText("Details")).toHaveLength(2);
  });
});
