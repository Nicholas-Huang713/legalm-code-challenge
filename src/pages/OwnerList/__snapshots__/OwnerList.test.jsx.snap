// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OwnerList Component > renders List component with correct props when ownerList is not empty 1`] = `
<div>
  <button>
    Adopt
  </button>
  <label
    for="filter"
  >
    Filter by Exp 
  </label>
  <select
    id="filter"
  >
    <option
      value="none"
    >
      None
    </option>
    <option
      value="1"
    >
      1
    </option>
    <option
      value="2"
    >
      2
    </option>
    <option
      value="3"
    >
      3
    </option>
    <option
      value="4"
    >
      4
    </option>
    <option
      value="5"
    >
      5
    </option>
    <option
      value="6"
    >
      6
    </option>
    <option
      value="7"
    >
      7
    </option>
    <option
      value="8"
    >
      8
    </option>
    <option
      value="9"
    >
      9
    </option>
    <option
      value="10"
    >
      10
    </option>
  </select>
  <div
    class="list-container"
  >
    <ul
      class="list-wrapper"
    >
      <li
        class="list-item"
      >
        <div
          class="list-content"
        >
          <div>
            <strong>
              John Doe
            </strong>
          </div>
          <div>
            Exp: 
            3
            yrs
          </div>
          <div>
            <button
              value="1"
            >
              X
            </button>
          </div>
        </div>
        <div
          class="list-btn-wrapper"
        >
          <button
            value="1"
          >
            Details
          </button>
          <button
            value="1"
          >
            Edit
          </button>
        </div>
      </li>
      <li
        class="list-item"
      >
        <div
          class="list-content"
        >
          <div>
            <strong>
              Jane Doe
            </strong>
          </div>
          <div>
            Exp: 
            5
            yrs
          </div>
          <div>
            <button
              value="2"
            >
              X
            </button>
          </div>
        </div>
        <div
          class="list-btn-wrapper"
        >
          <button
            value="2"
          >
            Details
          </button>
          <button
            value="2"
          >
            Edit
          </button>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
